Guava adapter

d-michail opened the pull request:
A Guava adapter module which allows users to use all algorithms of the JGraphT library using the Guava 'Network` implementations. 

Guava contains 3 graph interfaces (a) Graph, (b) ValueGraph, and (c) Network. The `Network` class is the more general one and corresponds almost 1-to-1 with our `Graph` interface.

Kept on-purpose separate from the ext module since Guava is somewhat known for its incompatibilities between versions.

jbduncan commented:
Do JGraphT `Graph`s need their edges to be unique in terms of value? (I.e., can more than one edge have the value "foo" if edges contain strings?)

If not, then Guava's `ValueGraph` would be more appropriate to map to instead of `Network`, because a ValueGraph doesn't require that its edges are unique - only Network does.

jbduncan commented:
Talking a bit off-topic, I've been wanting a feature like this for some time now, so I'm personally very excited for this PR! :)

d-michail commented on jgrapht-guava/src/main/java/org/jgrapht/graph/BaseNetworkAdapter.java:
```diff
@@ -0,0 +1,186 @@
+        if (unmodifiableEdgeSet == null) {
+            unmodifiableEdgeSet = Collections.unmodifiableSet(network.edges());
+        }
+        return unmodifiableEdgeSet;
```
In JGraphT all modifications of the edge set must be performed explicitly by calling methods add and remove edge. The adapter returns a JGraphT graph which behaves exactly like the other graph implementations.

Method 'edgesOf()' also returns an unmodifiable set.

jbduncan replied:
Cool, makes sense to me!

d-michail commented on jgrapht-guava/src/main/java/org/jgrapht/graph/BaseNetworkAdapter.java:
```diff
@@ -0,0 +1,186 @@
+ */
+public abstract class BaseNetworkAdapter<V, E, N extends Network<V, E>>
+    extends AbstractGraph<V, E>
+    implements Graph<V, E>, Cloneable, Serializable
```
Indeed. 

I had implemented custom serialization, but got lost somewhere between my PCs at home and work. I will check to find it.

d-michail replied:
I added the custom serialization. Thanks for pointing out.

jbduncan replied:
You're very welcome. :+1:

d-michail commented:
Regarding uniqueness, yes JGraphT's graph has the same requirements as the network in guava.

d-michail commented on jgrapht-guava/src/main/java/org/jgrapht/graph/BaseNetworkAdapter.java:
```diff
@@ -0,0 +1,186 @@
+    public E getEdge(V sourceVertex, V targetVertex)
+    {
+        return network
+            .edgesConnecting(sourceVertex, targetVertex).stream().findFirst().orElse(null);
```
This is a case of incompatibility between the two interfaces. Guava throws an exception if there are more than one edges (in case of a multi-graph), while JGraphT returns the first edge.

jbduncan replied:
Oh okay, that makes sense! Thanks for the clarification. :)

jbduncan commented:
> Regarding uniqueness, yes JGraphT's graph has the same requirements as the network in guava.

:+1:

jsichi commented:
Regarding packages, wouldn't it be better to create a new one?  It's a bit unusual for an optional module to inject new classes into an existing package (and might cause trouble down the road if we were ever to start sealing the core jar).

I'd be ok with org.jgrapht.guava, org.jgrapht.graph.guava, or org.jgrapht.guava.graph--whatever makes the most sense and leaves room open for expansion as we find other integration points.

jsichi commented:
Also, could you update README.md in this PR?  At a minimum, the Release Contents and Dependencies section.

jbduncan commented:
It occurs to me that at some point, this new module will need an automatic module name (assuming https://github.com/jgrapht/jgrapht/issues/484 is resolved in the foreseeable future). I've raised a comment over at https://github.com/jgrapht/jgrapht/issues/484 for future reference.

jsichi commented on jgrapht-guava/src/main/java/org/jgrapht/graph/WeightedGraphImmutableNetworkAdapter.java:
```diff
@@ -0,0 +1,162 @@
+import com.google.common.graph.NetworkBuilder;
+
+/**
+ * A weighted graph adapter class using Guava's {@link ImmutableNetwork}.
```
(And I just have to say, the guava interface split seems like they were intentionally trying to make it hard to write universally applicable algorithms...e.g. how would you write the equivalent of a ClosestFirstIterator?  The predecessor/successor functions aren't helpful since they don't provide any access to the edge information.  Well, I guess once we have a complete set of adapters, that will be the answer.)

d-michail replied:
The `ValueGraph` does not have any class for the edges, the first generic parameter is the type of the vertices and the second generic parameter is the type of the weight of the edges. 

Since our weight is `double` and not a generic parameter, this forces us to map to a graph with a custom edge class which contains the weight as a field. Ideas?

jbduncan replied:
@jsichi My understanding with the Guava `common.graph` interface split (as well as my personal experience with using both Guava and JGraphT) is that it makes things easier for users (rather than implementors or algorithm writers, necessarily), because e.g. if one doesn't need to associate info with edges, then they can just use `Graph` which has a by comparison a simpler API than `ValueGraph` and `Network`.

But you make an interesting point about `ClosestFirstIterator`. How _would_ one implement it for all 3 Guava types? :thinking:

jbduncan replied:
@d-michail How about introducing an adapter class that turns just a `ValueGraph<V, Double>` into a JGrapht `Graph`?

Alternatively, maybe we could introduce an adapter class that turns `ValueGraph<V, E>` into a JGraphT `Graph`, but the user has to pass in something like Guava's [`Converter`](https://static.javadoc.io/com.google.guava/guava/24.0-jre/com/google/common/base/Converter.html) to the constructor, where `converter.convert()` would turn a given edge of type `E` into a weight of type `Double`, and `converter.reverse().convert()` would do the opposite?

jsichi replied:
Since @d-michail removed the weighted adapter in favor of AsWeightedGraph for now, I'll restart this conversation at top level as notes for a followup pull request.

jbduncan replied:
@jsichi Oops, I'd forgotten that @d-michail wrote that comment! Indeed, let's restart the conversation at top level. :)

d-michail commented:
@jsichi Indeed, I will create a new package. I already faced a few problems with javadoc due to the injection.

jsichi commented on jgrapht-guava/src/main/java/org/jgrapht/graph/GraphImmutableNetworkAdapter.java:
```diff
@@ -0,0 +1,188 @@
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ */
+public class GraphImmutableNetworkAdapter<V, E>
```
For the immutable variants, it seems like it would be easier to just our our AsUnmodifiable wrapper, but I haven't thought it through, so if there's an obvious reason not to, just ignore this comment.

d-michail commented on jgrapht-guava/src/main/java/org/jgrapht/graph/WeightedGraphImmutableNetworkAdapter.java:
```diff
@@ -0,0 +1,162 @@
+        return newGraph;
+    }
+    
+    private void writeObject(ObjectOutputStream oos)
```
Live and learn! Thanks

d-michail commented on jgrapht-guava/src/test/java/org/jgrapht/graph/GraphImmutableNetworkAdapterTest.java:
```diff
@@ -0,0 +1,264 @@
+    }
+   
+
+    private Object serializeAndDeserialize(Object obj)
```
I just left one copy, to avoid having test package dependencies.

jsichi replied:
Is it bad to have test package dependencies?  We should probably have a place for common test-only utilities.  But I'm fine with leaving that refactoring for a followup pull request.

d-michail commented:
I recalled why I am using the same package (org.jgrapht.graph). It is due to the fact that we are using an intrusive edge (which is package private) in the weighted variant.

One solution would be to completely skip the weighted variants, and let users use the 'AsWeighted` view.

d-michail commented:
@jsichi  Removed completely the weighted variant (can be done with `AsWeightedGraph`), and renamed the package.

jsichi commented on README.md:
```diff
@@ -51,15 +51,22 @@ The files below make up the table of contents for a release distribution archive
+    - `animal-shiffer-annotations-x.y.jar` Guava dependency
+    - `checker-compat-qual-x.y.z.jar` Guava dependency
+    - `error_prone_annotations-x.y.z.jar` Guava dependency
+    - `jsr305-x.y.z.jar` Guava dependency
```
I ran mvn package on your branch, and I see one more file in the distribution tarball:  j2objc-annotations-1.1.jar.

jsichi commented:
Another followup note:

https://github.com/jgrapht/jgrapht/issues/534

jbduncan commented:
@jsichi 

> I think the most natural way to do this would be to use EndpointPair directly as our edge type

Agreed!

> To implement getWeight (in the case where the user requests that the double value be interpreted as a weight), we would just call edgeValue on u/v from the EndpointPair.

I can't speak for @d-michail, but I'm not sure I understand yet what you mean here. AFAICT, there still needs to be a way for the u/v pair from an EndpointPair to be turned into a `double` weight, and I think it would have to fall on the user to provide this as a custom `ToDoubleBiFunction<U, V>` that gets passed as a parameter to the adapter's constructor.

Am I missing something here? :)

> getEdgeSource/getEdgeTarget can likewise be implemented by interrogating the EndpointPair.

I agree with this as well!

jbduncan commented:
@jsichi Oh yep, you're right! For the given adapter class, which operates solely on `ValueGraph<V, Double>`s rather than more generic `ValueGraph<V, E>`s, then AFAICT that implementation would indeed suffice. Thanks for the clarification!

> But as you suggest, a converter might be desirable for more flexibility in the case where the value is actually a complex object.

Acknowledged and agreed. But thinking a bit more about it, I wonder just _how_ desirable it would be to pass in such a converter function. Would it be worth waiting until a user asks for it?

d-michail commented:
Thanks for the ideas. 

One more question about this. Since we are going to use `EndpointPair<V>` as our edge, how does method `addEdge(V sourceVertex, V targetVertex, Endpoint<V> e) behave?

 - either we completely ignore the passed parameter e
 - or we keep a Map<Pair<V,V>,EndpointPair<V>>

I would personally prefer to do the first and simply state in the Javadocs that the explicitly given object is silently ignored. Otherwise the adapter will become an graph implementation and not a delegator.

d-michail commented:
Added adapters for both `ValueGraph` and `Graph`. For value graph I added a general one and a specialized one for double values.

- MutableGraphAdapter maps `Graph<V>` to `Graph<V, EndpointPair<V>>` 
- MutableValueGraphAdapter maps `ValueGraph<V,W>` to `Graph<V,EndpointPair<V>>` and requires the user to provide a converter method W->double and a default value for W.  The default value is needed in order to be able to implement `addEdge()` on the adapter. Regarding the weights, this is one-directional, it does not implement `setEdgeWeight()`.  I could add another lambda on the constructor for performing the reverse, updating a value when a user calls `setEdgeWeight()`. Does this make sense?
- MutableDoubleValueGraphAdapter maps `ValueGraph<V,Double>` to `Graph<V,EndpointPair<V>>`. Does not require any converter and `setEdgeWeight()` works as expected.

Each Mutable has a similar Immutable variant.

d-michail commented:
@jsichi Did several changes in packaging which affect #534. Take a look and tell me if we need something more. 

I added the licenses in META-INF/ inside the uber jars using shade-plugin and also added the licenses in the main distribution using the assembly plugin.

I also refactored a little bit the relative paths (mainly defined the property main.basedir) for the parent's basedir. I do not think so, but please double check that this will not break something during the next release.

jsichi commented on README.md:
```diff
@@ -51,15 +51,23 @@ The files below make up the table of contents for a release distribution archive
+    - `antlr4-runtime-x.y.jar` ANTLR parser runtime
+    - `commons-lang3-x.y.jar` Apache Commons Lang library
+    - `guava-x.y-jre.jar` Guava library
+    - `animal-shiffer-annotations-x.y.jar` Guava dependency
```
Heh, it's actually animal-sniffer.  (Ew!)  And somehow it turned into animal-shifter further down...lycanthropy perhaps?

jsichi replied:
Also, I'm wondering whether we can actually exclude all the annotations dependencies to keep things more streamlined.  It wasn't clear from here:

https://github.com/google/guava/issues/2824

jsichi commented on jgrapht-guava/pom.xml:
```diff
@@ -0,0 +1,128 @@
+						</transformer>
+						<transformer implementation="org.apache.maven.plugins.shade.resource.IncludeResourceTransformer">
+							<resource>META-INF/animal-shifter-annotations-license</resource>
+							<file>${main.basedir}/etc/licenses/MIT-license.txt</file>
```
That copy of the license leaves the copyright line holder and year blank.  I think the best we can do is to use an extract from the pom.xml:

https://github.com/mojohaus/animal-sniffer/blob/master/animal-sniffer-annotations/pom.xml

d-michail commented on jgrapht-guava/src/main/java/org/jgrapht/graph/guava/MutableValueGraphAdapter.java:
```diff
@@ -0,0 +1,269 @@
+        MutableValueGraph<V, W> valueGraph, W defaultValue, ToDoubleFunction<W> valueConverter)
+    {
+        super(valueGraph, valueConverter);
+        this.defaultValue = Objects.requireNonNull(defaultValue);
```
That was also my initial thought, until I realized that there is no `addEdge` method!

jbduncan replied:
@jsichi My understanding is that `ValueGraph.edgeValue(u, v)` returns a non-empty Optional if there is an edge between u and v, otherwise `Optional.empty()`. (That is, if there is no edge between u and v, it doesn't throw an exception; it just returns `Optional.empty()` instead.)

The reason I think this is because `MutableValueGraph` has only one method that adds edges: `putEdgeValue`, and AFAICT one has to provide an edge object to that method, in turn meaning that every edge in a `ValueGraph` has an associated edge object.

I could be wrong about this, though. :)

jbduncan commented on jgrapht-guava/src/main/java/org/jgrapht/graph/guava/BaseGraphAdapter.java:
```diff
@@ -0,0 +1,214 @@
+    {
+        return graph
+            .predecessors(vertex).stream().map(other -> createEdge(other, vertex))
+            .collect(Collectors.toSet());
```
Considering that:

1. The current Guava graph implementations generally return unmodifiable views for methods that return `Set`s (according to https://github.com/google/guava/wiki/GraphsExplained#accessor-behavior), and
2. `Collectors.toSet()` returns a mutable set as of Java 8 and that users may expect that if they modify this set, then the underlying graph will also be changed

...I think we should swap out this line of code for this one instead:

```java
    .collect(collectingAndThen(toSet(), Collections::unmodifiableSet));
```

jbduncan commented on jgrapht-guava/src/main/java/org/jgrapht/graph/guava/BaseGraphAdapter.java:
```diff
@@ -0,0 +1,214 @@
+    {
+        return graph
+            .successors(vertex).stream().map(other -> createEdge(vertex, other))
+            .collect(Collectors.toSet());
```
Ditto.

jbduncan commented on jgrapht-guava/src/main/java/org/jgrapht/graph/guava/BaseValueGraphAdapter.java:
```diff
@@ -0,0 +1,222 @@
+    {
+        return valueGraph
+            .predecessors(vertex).stream().map(other -> createEdge(other, vertex))
+            .collect(Collectors.toSet());
```
Ditto.

jbduncan commented on jgrapht-guava/src/main/java/org/jgrapht/graph/guava/BaseValueGraphAdapter.java:
```diff
@@ -0,0 +1,222 @@
+    {
+        return valueGraph
+            .successors(vertex).stream().map(other -> createEdge(vertex, other))
+            .collect(Collectors.toSet());
```
Ditto.

jbduncan commented:
@jsichi 

> Also, are we going to send an emissary over to Guava and let them know about these? Hopefully the messenger won't return with a directed edge sticking out of his back :)

;)

I wonder if @jrtom, the lead maintainer of Guava `common.graph`, would be interested to give feedback on this?

jsichi commented:
@d-michail I added the exclusions below (adapted from https://github.com/t-tsutsumi/dropwizard-bom/commit/6aade001fcc04299545b128f8bc3982110fbdf71) inside the dependency in jgrapht-guava/pom.xml, and build and test ran through fine.  Can we use this to streamline the dependencies?

```
      <exclusions>
                    <exclusion>
                        <groupId>com.google.code.findbugs</groupId>
                        <artifactId>jsr305</artifactId>
                    </exclusion>
                    <exclusion>
                        <groupId>com.google.errorprone</groupId>
                        <artifactId>error_prone_annotations</artifactId>
                    </exclusion>
                    <exclusion>
                        <groupId>com.google.j2objc</groupId>
                        <artifactId>j2objc-annotations</artifactId>
                    </exclusion>
                    <exclusion>
                        <groupId>org.codehaus.mojo</groupId>
                        <artifactId>animal-sniffer-annotations</artifactId>
                    </exclusion>
                    <exclusion>
                        <groupId>org.checkerframework</groupId>
                        <artifactId>checker-compat-qual</artifactId>
                    </exclusion>
      </exclusions>
```

jsichi commented:
@jrtom Nice to meet you too, Joshua!  We're international (in fact none of us have met each other in person either, but we're hoping to fix that soon); I'm probably the closest to you (in San Francisco).  So yeah, let's grab a beer next time we're in the same city.

Dimitrios considered putting this adapter into a separate project, but that has its own development+release overhead, so for now we're starting it as an optional module of the main project (so that we can keep our core free of dependencies on Guava like you mention).  If things get sticky in the future (due to either out-of-sync release cycles or a desire for a matrix of version combinations), we can move it out at that time.

d-michail commented:
@jsichi Added the exclusion as you suggested and removed the additional licenses from the package. If someone complains downstream we should probably suggest them to exclude the guava dependency from  us and add it separately on their project.

d-michail commented:
@jrtom Nice to meet you Joshua. Hopefully, this year's GSoC will be an opportunity to meet in person with John (@jsichi) and Joris (@jkinable).

jsichi commented:
@d-michail The licenses for the excluded dependencies are still being written to the jgrapht-guava-uber jar...I guess we don't need them there any more.

@jkinable any comments on this PR?  If not, I'm planning to merge it after this last remaining item gets resolved.

jkinable commented on jgrapht-guava/src/main/java/org/jgrapht/graph/guava/MutableValueGraphAdapter.java:
```diff
@@ -0,0 +1,270 @@
+ * The class uses a converter from Guava's values to JGraphT's double weights. Thus, the resulting
+ * graph is weighted. This is a one-way conversion meaning that calling
+ * {@link #setEdgeWeight(EndpointPair, double)} will throw an unsupported operation exception.
+ * Adjusting the weights can be done directly (by keeping an external reference) on the underlying
```
can you be a little more explicit here (perhaps a small example)? I'm not sure that it is very obvious how to change the weight of an edge.

jsichi commented:
Thanks, nice work!
